env/env_inspected.cc

```diff
@@ -0,0 +1,349 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+#ifndef ROCKSDB_LITE
+#include "rocksdb/env_inspected.h"
+
+namespace rocksdb {
+
+class InspectedSequentialFile : public SequentialFileWrapper {
+ public:
+  InspectedSequentialFile(std::unique_ptr<SequentialFile>&& target,
+                          std::shared_ptr<FileSystemInspector> inspector)
+      : SequentialFileWrapper(target.get()),
+        owner_(std::move(target)),
+        inspector_(inspector) {}
+
+  Status Read(size_t n, Slice* result, char* scratch) override {
+    assert(inspector_);
+    Status s;
+    size_t offset = 0;
+    size_t allowed = 0;
+    while (offset < n) {
+      s = inspector_->Read(n - offset, allowed);
+      if (!s.ok()) {
+        return s;
+      }
+      assert(allowed <= n - offset);
+      if (allowed > 0) {
+        s = SequentialFileWrapper::Read(allowed, result, scratch + offset);
```

Comment:
should also use positionedread

Reply:
and seems all the logic of Read() of the different classes are same, can you extract the common logic?

Reply:
Nah, Read will update file pointer offset internally, plus we don't really know the right offset to pass to PositionedRead.

Reply:
oh, got it